Mathlib/Tactic/NormNum/NormNumI.lean

```diff
@@ -0,0 +1,222 @@
+
+theorem IsComplex.re_eq {z : ℂ} {a b : ℝ} (hz : IsComplex z a b) : Complex.re z = a := by
+  simp [hz.out]
+
+theorem IsComplex.im_eq {z : ℂ} {a b : ℝ} (hz : IsComplex z a b) : Complex.im z = b := by
+  simp [hz.out]
+
+theorem IsComplex.of_pow_negSucc {z w : ℂ} {a b : ℝ} {n : ℕ} {k k' : ℤ} (h : k = Int.negSucc n)
+    (hk : NormNum.IsInt k' k)
+    (hz : IsComplex (w ^ (n + 1))⁻¹ a b) (hz' : z = w ^ (k' : ℤ)) : IsComplex z a b := by
+  obtain rfl : k' = k := by simpa using hk.out
+  constructor
+  simpa [h, hz'] using hz.out
+
+theorem IsComplex.of_pow_ofNat {z w : ℂ} {k k' : ℤ} {n : ℕ} {a b : ℝ} (hz1 : z = w ^ k)
+    (hw : IsComplex (w ^ n) a b) (hk' : k' = n) (hkk' : NormNum.IsInt k k') :
+    IsComplex z a b := by
+  obtain rfl : k = k' := by simpa using hkk'.out
+  constructor
+  simpa [hz1, hk'] using hw.out
+
+/-- Parsing all the basic calculation in complex. -/
+partial def parse (z : Q(ℂ)) : MetaM (Σ a b : Q(ℝ), Q(IsComplex $z $a $b)) := do
+  match z with
+  | ~q($z₁ + $z₂) =>
+    let ⟨_a₁, _b₁, pf₁⟩ ← parse z₁
+    let ⟨_a₂, _b₂, pf₂⟩ ← parse z₂
+    pure ⟨_, _, q(.add $pf₁ $pf₂)⟩
+  | ~q($z₁ * $z₂) =>
+    let ⟨_a₁, _b₁, pf₁⟩ ← parse z₁
+    let ⟨_a₂, _b₂, pf₂⟩ ← parse z₂
+    pure ⟨_, _, q(.mul $pf₁ $pf₂)⟩
+  | ~q($z⁻¹) =>
+    let ⟨_x, _y, pf⟩ ← parse z
+    pure ⟨_, _, q(.inv $pf)⟩
+  | ~q($z₁ / $z₂) => do
+    let ⟨_a, _b, pf⟩ ← parse q($z₁ * $z₂⁻¹)
+    return ⟨_, _, q($pf)⟩
+  | ~q(-$w) => do
+    let ⟨_a, _b, pf⟩ ← parse w
+    return ⟨_, _, q(.neg $pf)⟩
+  | ~q($z₁ - $z₂) =>
+    let ⟨_a₁, _b₁, pf₁⟩ ← parse z₁
+    let ⟨_a₂, _b₂, pf₂⟩ ← parse z₂
+    pure ⟨_, _, q(.sub $pf₁ $pf₂)⟩
+  | ~q(conj $w) =>
+    let ⟨_a, _b, pf⟩ ← parse w
+    return ⟨_, _, q(.conj $pf)⟩
+  | ~q($w ^ ($n' : ℕ)) =>
+    let ⟨n, hn⟩ ← NormNum.deriveNat q($n') q(inferInstance)
+    match n.natLit! with
+    | 0 =>
+      let _i : $n =Q 0 := ⟨⟩
+      return ⟨q(1), q(0), q(⟨show $w ^ $n' = _ from $(hn).out ▸ pow_zero _⟩)⟩
+    | n + 1 =>
+      parse q($w^$n * $w)
+  | ~q($w ^ ($k : ℤ)) =>
+    let ⟨k', hm⟩ ← NormNum.deriveInt q($k) q(inferInstance)
+    match k'.intLit! with
+    | Int.ofNat n =>
+      let ⟨a, b, pf⟩ ← parse q(@HPow.hPow ℂ ℕ ℂ instHPow $w $n)
+      let _i : $k' =Q $n := ⟨⟩
+      return ⟨a, b, q(.of_pow_ofNat rfl $pf rfl $hm)⟩
+    | Int.negSucc n =>
+      let ⟨a, b, pf⟩ ← parse q(($w ^ ($n + 1))⁻¹)
+      let _i : $k' =Q Int.negSucc $n := ⟨⟩
+      return ⟨a, b, q(.of_pow_negSucc rfl $hm $pf rfl)⟩
+  | ~q(Complex.I) =>
+    pure ⟨_, _, q(.I)⟩
+  | ~q(0) =>
+    pure ⟨_, _, q(.zero)⟩
+  | ~q(1) =>
+    pure ⟨_, _, q(.one)⟩
+  | ~q(OfNat.ofNat $en (self := @instOfNatAtLeastTwo ℂ _ _ $inst)) =>
+    return ⟨_, _, q(.ofNat $en)⟩
+  | ~q(OfScientific.ofScientific $m $x $exp) =>
+    return ⟨_, _, q(.scientific _ _ _)⟩
+  | _ => throwError "found the atom {z} which is not a numeral"
+
+/-- Using `norm_num` to normalise expressions -/
+def normalize (z : Q(ℂ)) : MetaM (Σ a b : Q(ℝ), Q(IsComplex $z $a $b)) := do
```

Comment:
I think this function should actually be
```suggestion
def normalize (z : Q(ℂ)) : MetaM (Simp.ResultQ q($z)) := do
```
and should look at `ra` and `rb` to decide something like
* `$a`
* `-$(-b)*I`
* `-I`
* `I`
* `$b*I`
* `$a - $b*I`
* `$a - I`
* `$a + I`
* `$a + $b*I`

depending on the sign of `b` and whether it is `1`/`0`/`-1`.

Reply:
I agree with this suggestion, if what you mean is (more precisely)
```lean
def normalizeAux (z : Q(ℂ)) : MetaM (Σ a b : Q(ℝ), Q(IsComplex $z $a $b)) := do
  -- this is a renaming of the existing `normalize` function

def normalize (z : Q(ℂ)) : MetaM (Simp.ResultQ q($z)) := do
  -- lightweight wrapper for `normalizeAux`?
```
Is that indeed what you mean?